dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/EventReferralValidationHookTest.java

```diff
@@ -0,0 +1,254 @@
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 0, reporter.getReportList().size() );
+    }
+
+    @Test
+    public void verifyValidationSuccessIfReferralHasCompleteRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .relationships( Arrays.asList( relationship ) )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 0, reporter.getReportList().size() );
+    }
+
+    @Test
+    public void verifyValidationFailIfReferralHasInCompleteRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .relationships( Arrays.asList( relationshipIncomplete ) )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 1, reporter.getReportList().size() );
+        assertTrue( reporter.hasErrorReport( r -> r.getErrorCode() == TrackerErrorCode.E1312 ) );
+    }
+
+    @Test
+    public void verifyValidationFailIfReferralHasNoRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 1, reporter.getReportList().size() );
+        assertTrue( reporter.hasErrorReport( r -> r.getErrorCode() == TrackerErrorCode.E1311 ) );
+    }
+
+    private void init()
+    {
+        orgUnit = createOrganisationUnit( "A" );
```

Comment:
Do you really need to initialise all these metadata for the test?
Unnecessary details make it difficult to understand what is being tested and what is important for the test.
I would suggest to remove any metadata that will not change the result of the test

Reply:
resolved